pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -1509,6 +1514,31 @@ public CompletableFuture<List<String>> getListOfTopics(NamespaceName namespaceNa
         }
     }
 
+    public CompletableFuture<List<String>> getListOfUserTopics(NamespaceName namespaceName, Mode mode) {
+        String key = String.format("%s://%s", mode, namespaceName);
+        CompletableFuture<List<String>> queryRes =
+                inProgressQueryUserTopics.computeIfAbsent(key, k -> {
+            CompletableFuture<List<String>> res = new CompletableFuture<>();
+            // Switch thread to avoid blocking other threads who are calling the current method.
+            pulsar.getExecutor().execute(() -> {
+                getListOfTopics(namespaceName, mode).thenApply(list -> {
+                    return TopicList.filterSystemTopic(list);
+                }).whenComplete((topics, ex) -> {
+                    if (ex != null) {
+                        res.completeExceptionally(ex);
+                    } else {
+                        res.complete(topics);
+                    }
+                });
+            });
+            return res;
+        });
+        queryRes.whenComplete((ignore, ex) -> {
```

Comment:
A best practice is to check if the `computeIfAbsent` inserts a new entry. Otherwise, if there are N concurrent `getListOfUserTopics` calls, the same future will register N `whenComplete` callbacks.

Reply:
The existing code looks like:

```java
    static class Cache {
        private final Map<Integer, CompletableFuture<Integer>> futures = new ConcurrentHashMap<>();

        public CompletableFuture<Integer> run(int x) {
            final var future = futures.computeIfAbsent(x, __ -> new CompletableFuture<>());
            future.whenComplete((__, ___) -> {
                System.out.println("Remove " + x);
                futures.remove(x, future);
            });
            return future;
        }
    }
```

Given the following code:

```
        final var cache = new Cache();
        cache.run(1);
        cache.run(1);
        cache.run(1);
        cache.run(1).complete(100);
```

Outputs:

```
Remove 1
Remove 1
Remove 1
Remove 1
```

The fix:

```java
        public CompletableFuture<Integer> run(int x) {
            final var inserted = new MutableBoolean(false);
            final var future = futures.computeIfAbsent(x, __ -> {
                inserted.setTrue();
                return new CompletableFuture<>();
            });
            if (inserted.isTrue()) {
                future.whenComplete((__, ___) -> {
                    System.out.println("Remove " + x);
                    futures.remove(x, future);
                });
            }
            return future;
        }
```

Reply:
I think they all use the same future is better 😂

Reply:
Checking if it's inserted does not change the fact that all concurrent `run` calls return the same future

Reply:
> Outputs:
> Remove 1
> Remove 1
> Remove 1
> Remove 1

The code you provided is the evidence that the first task blocks other threads, which answered your below question: https://github.com/apache/pulsar/pull/23049#discussion_r1682326069

>The existing code looks like:
> ...
> The fix:
> ...

The difference between the two examples: the `CompletableFuture` in the second code never complete

Reply:
I think I got what you want to say, I fixed it

Reply:
> the CompletableFuture in the second code never complete.

See the following code and outputs.

```java
    static class Cache {
        private final Map<Integer, CompletableFuture<Integer>> futures = new ConcurrentHashMap<>();

        public CompletableFuture<Integer> run(int x) {
            final var inserted = new MutableBoolean(false);
            final var future = futures.computeIfAbsent(x, __ -> {
                inserted.setTrue();
                final var innerFuture = new CompletableFuture<Integer>();
                CompletableFuture.runAsync(() -> {
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                    }
                    innerFuture.complete(100);
                });
                return innerFuture;
            });
            if (inserted.isTrue()) {
                future.whenComplete((__, ___) -> {
                    System.out.println("Remove " + x);
                    futures.remove(x, future);
                });
            }
            return future;
        }
    }

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        final var cache = new Cache();
        cache.run(1).thenAccept(System.out::println);
        cache.run(1).thenAccept(System.out::println);
        cache.run(1).thenAccept(System.out::println);
        cache.run(1).thenAccept(System.out::println).get();
```

Outputs:

```
100
100
100
100
Remove 1
```

Reply:
`CompletableFuture.runAsync` is the same as `executor.execute(runnable)`, you switched the thread too